# Stduent Performance Indicator

## Life cycle of Machine Learning Project:

* Understanding teh problem statment
* Data Collection
* Data checks to perform
* Explotary data analyst (eda)
* Data Per-processing
* Model Training
* Choose the best model

## How to understand the steps: 
* Step 1: the question. **Decide what question you're answering.**
* Step 2: get data. Do we actually have data that can answer this problem? (if so, Get the information you'll learn from.)
* Steps 3: validate data. **Can I trust this data?**

### 1) **Problem statement**
* This project understands how the student's performance (test scores) is affected by other variables such as Gender, Ethnicity, Parental level of education, Lunch and Test preparation course.

### 2) **Data Collection:** 
* Download the data from kaggle and put it in ``notebooks/data``. Provice the number of column and rows: 
* The data consist of 8 colomns and 1000 rows. 

#### 2.1) Import the data and required packages. 
* Import ``Pandas, Numpy, Matplotlib, Seaborn`` and Warning libraries.  

In [10]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

#### 2.2) Import the CSV Data as Pandas DataFrame

In [11]:
df = pd.read_csv('data/stud.csv')

# Showing the top 5 records: 
df.head()

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score
0,female,group B,bachelor's degree,standard,none,72,72,74
1,female,group C,some college,standard,completed,69,90,88
2,female,group B,master's degree,standard,none,90,95,93
3,male,group A,associate's degree,free/reduced,none,47,57,44
4,male,group C,some college,standard,none,76,78,75


#### 2.3) Check the shape of the data:

In [7]:
df.shape

(1000, 8)

#### 2.5) Data Information: 
* gender : sex of students -> (Male/female)
* race/ethnicity : ethnicity of students -> (Group A, B, C, D,E)
* parental level of education : parents' final education ->(bachelor's degree,some college,master's degree,associate's degree,high school)
* lunch : having lunch before test (standard or free/reduced)
* test preparation course : complete or not complete before test
* math score
* reading score
* writing score


### 3) **Data Checks to Perform**
* Check Missing values
* Check Duplicates
* Check data type
* Check the number of unique values of each column
* Check statistics of data set
* Check various categories present in the different categorical column

#### 3.1) Check Missing Values: 
If many values are missing, you'll need to decide whether to:
1. remove rows,
2. fill in missing values,
3. or collect better data.


In [ ]:
# 3.1 Check number of missing values:
df.isna().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

In [13]:
# 3.2 Check Duplicates:
df.duplicated().sum()

np.int64(0)

In [15]:
# 3.3 Check data types
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype
---  ------                       --------------  -----
 0   gender                       1000 non-null   str  
 1   race/ethnicity               1000 non-null   str  
 2   parental level of education  1000 non-null   str  
 3   lunch                        1000 non-null   str  
 4   test preparation course      1000 non-null   str  
 5   math score                   1000 non-null   int64
 6   reading score                1000 non-null   int64
 7   writing score                1000 non-null   int64
dtypes: int64(3), str(5)
memory usage: 62.6 KB


In [16]:
# 3.4 Check the number of unique values in each colunm:
df.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     81
reading score                  72
writing score                  77
dtype: int64

In [17]:
# 3.5 Check Statics of Data
df.describe()

,math score,reading score,writing score
count,1000.00000,1000.000000,1000.000000
mean,66.08900,69.169000,68.054000
std,15.16308,14.600192,15.195657
min,0.00000,17.000000,10.000000
25%,57.00000,59.000000,57.750000
50%,66.00000,70.000000,69.000000
75%,77.00000,79.000000,79.000000
max,100.00000,100.000000,100.000000


**Insight**

* From above description of numerical data, all means are very close to each other - between 66 and 68.05;
* All standard deviations are also close - between 14.6 and 15.19;
* While there is a minimum score 0 for math, for writing minimum is much higher = 10 and for reading myet higher = 17

In [18]:
# 3.6 Exploring the data:
df.head()

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score
0,female,group B,bachelor's degree,standard,none,72,72,74
1,female,group C,some college,standard,completed,69,90,88
2,female,group B,master's degree,standard,none,90,95,93
3,male,group A,associate's degree,free/reduced,none,47,57,44
4,male,group C,some college,standard,none,76,78,75


In [22]:
print("Categories in 'gender' variable:     ",end=" " )
print(df['gender'].unique())
print()
print("Categories in 'race_ethnicity' variable:  ",end=" ")
print(df['race/ethnicity'].unique())
print()
print("Categories in'parental level of education' variable:",end=" " )
print(df['parental level of education'].unique())
print()
print("Categories in 'lunch' variable:     ",end=" " )
print(df['lunch'].unique())
print()
print("Categories in 'test preparation course' variable:     ",end=" " )
print(df['test preparation course'].unique())

Categories in 'gender' variable:      <StringArray>
['female', 'male']
Length: 2, dtype: str

Categories in 'race_ethnicity' variable:   <StringArray>
['group B', 'group C', 'group A', 'group D', 'group E']
Length: 5, dtype: str

Categories in'parental level of education' variable: <StringArray>
[ 'bachelor's degree',       'some college',    'master's degree',
 'associate's degree',        'high school',   'some high school']
Length: 6, dtype: str

Categories in 'lunch' variable:      <StringArray>
['standard', 'free/reduced']
Length: 2, dtype: str

Categories in 'test preparation course' variable:      <StringArray>
['none', 'completed']
Length: 2, dtype: str


In [25]:
# define numerical & categorical columns
numeric_features = [feature for feature in df.columns if df[feature].dtype == 'int64']
categorical_features = [feature for feature in df.columns if df[feature].dtype != 'int64']

# print columns
print(df["math score"].dtype)
print('We have {} numerical features : {}'.format(len(numeric_features), numeric_features))
print('\nWe have {} categorical features : {}'.format(len(categorical_features), categorical_features))

int64
We have 3 numerical features : ['math score', 'reading score', 'writing score']

We have 5 categorical features : ['gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course']


In [26]:
# 3.8 Adding columns for "Total Score" and "Average"
df["Total Score"] = df["math score"] + df["writing score"] + df["reading score"]
df["Average of Scores"] = df["Total Score"] / 3
df.head()

,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score,Total Score,Average of Scores
0,female,group B,bachelor's degree,standard,none,72,72,74,218,72.666667
1,female,group C,some college,standard,completed,69,90,88,247,82.333333
2,female,group B,master's degree,standard,none,90,95,93,278,92.666667
3,male,group A,associate's degree,free/reduced,none,47,57,44,148,49.333333
4,male,group C,some college,standard,none,76,78,75,229,76.333333


In [ ]:
reading_full = df[df["reading score"] == 100]["Average of Scores"].count()
math_full = df[df["math score"] == 100]["Average of Scores"].count()
writing_full = df[df["writing score"] == 100]["Average of Scores"].count()

print(f'Number of students with full marks in Maths: {math_full}')
print(f'Number of students with full marks in Writing: {writing_full}')
print(f'Number of students with full marks in Reading: {reading_full}')

Number of students with full marks in Maths: 7
Number of students with full marks in Writing: 14
Number of students with full marks in Reading: 17


In [35]:
reading_less_20 = df[df['reading score'] <= 20]['Average of Scores'].count()
writing_less_20 = df[df['writing score'] <= 20]['Average of Scores'].count()
math_less_20 = df[df['math score'] <= 20]['Average of Scores'].count()

print(f'Number of students with less than 20 marks in Maths: {math_less_20}')
print(f'Number of students with less than 20 marks in Writing: {writing_less_20}')
print(f'Number of students with less than 20 marks in Reading: {reading_less_20}')

Number of students with less than 20 marks in Maths: 4
Number of students with less than 20 marks in Writing: 3
Number of students with less than 20 marks in Reading: 1


**Insights**
* From above values we get students have performed the worst in Maths
* Best performance is in reading section

### 4) **Exploring Data ( Visualization )**